source/common/upstream/upstream_impl.cc

```diff
@@ -60,6 +60,8 @@ namespace Envoy {
 namespace Upstream {
 namespace {
 
+SINGLETON_MANAGER_REGISTRATION(const_common_lb_config_shared_pool);
```

Comment:
drive-by: can this be done in the cluster-manager or some other structure where there's one of them, but is not a singleton?

I'd like to avoid proliferating a lot of singletons if possible, in light of a few things like the fact like unit tests, environments like  Envoy/Mobile where multiple unrelated Envoy instances and be co-resident in a process with separate Enovy::Server, separate main-thread, etc.

Reply:
I'll double-down on this comment. This is not a good pattern generally IMO and I would push back in any case. But there's a specific problem with this one.

From source/common/shared_pool/shared_pool.h :

> Note:  ObjectSharedPool needs to be created in the main thread, all the member methods can only be called in the main thread, ...

And in Envoy-Mobile there are multiple main threads. So you really can't share ObjectSharedPool via a singleton.

Reply:
From my reading of the code `Singleton::Manager`s are not true (process wide) singletons and are unique only to the Envoy instance: https://github.com/envoyproxy/envoy/blob/main/source/server/server.h#L366

The `Singleton::Manager` itself does not have any static member objects.

Reply:
I think that works fine for serially instantiated and destructed server objects.

But that's not what happens in E-M. In that context there are multiple servers co-existing, each with their own main threads.

So being in *a* "main thread" does not mean you are in *the* main thread, and the ObjectStore (IIUC) makes that assumption.